File: ios/Tuner/Metronome/MetronomeScreen.swift
import SwiftUI

/// 节拍器面板（design-system §6.7）。
struct MetronomeView: View {
    @Environment(\.lumen) private var palette
    @StateObject private var vm = MetronomeViewModel()

    var body: some View {
        AuroraBackground(tuneCents: nil) {
            VStack(spacing: Lumen.Spacing.md) {
                // BPM 环 + 垂直拨轮
                BpmRing(bpm: vm.bpm) { delta in
                    vm.adjustBpm(delta.rounded())
                }
                .frame(height: 220)
                .overlay(alignment: .center) {
                    VStack {
                        Text("\(Int(vm.bpm))")
                            .font(Lumen.displayBpm)
                            .foregroundStyle(palette.accent)
                        Text("BPM（上下拖动调节）")
                            .font(Lumen.caption)
                            .foregroundStyle(palette.inkSecondary)
                    }
                }

                // 摆锤
                Pendulum(playing: vm.playing, bpm: vm.bpm, beatUnit: vm.beatUnit)
                    .frame(height: 96)

                // 滑杆 + 步进 + tap
                HStack {
                    StepButton("-5") { vm.adjustBpm(-5) }
                    StepButton("-1") { vm.adjustBpm(-1) }
                    Slider(value: Binding(
                        get: { vm.bpm },
                        set: { vm.setBpm($0) }
                    ), in: 30...250)
                    StepButton("+1") { vm.adjustBpm(1) }
                    StepButton("+5") { vm.adjustBpm(5) }
                }
                Button("TAP 测速") { vm.tap() }
                    .font(Lumen.label)
                    .foregroundStyle(palette.accent)
                    .padding(.horizontal, 16)
                    .padding(.vertical, 8)
                    .background(palette.bgSurface, in: Capsule())

                // 拍号选择
                ScrollView(.horizontal, showsIndicators: false) {
                    HStack(spacing: 6) {
                        ForEach(MetronomeViewModel.commonTimeSignatures, id: \.0) { sig in
                            let selected = vm.beatsPerBar == sig.0 && vm.beatUnit == sig.1
                            Button { vm.setTimeSignature(beats: sig.0, unit: sig.1) } label: {
                                Text("\(sig.0)/\(sig.1)")
                                    .font(Lumen.label)
                                    .foregroundStyle(selected ? palette.bgCanvas : palette.inkPrimary)
                                    .padding(.horizontal, 12)
                                    .padding(.vertical, 8)
                                    .background(selected ? palette.accent : palette.bgSurface, in: Capsule())
                            }
                        }
                    }
                }

                // 重音圆点行（重拍实心/普通半透/静音空心）
                HStack(spacing: 12) {
                    ForEach(vm.accents.indices, id: \.self) { i in
                        let isCurrent = vm.playing && vm.currentBeat == i
                        BeatDot(
                            accent: vm.accents[i],
                            current: isCurrent
                        ) { vm.cycleAccent(i) }
                    }
                }
                // 小节进度
                ProgressView(value: vm.playing && vm.currentBeat >= 0
                    ? Double(vm.currentBeat + 1) / Double(vm.beatsPerBar)
                    : 0)
                    .tint(palette.accent)

                // 音色选择（重拍/弱拍）
                HStack(spacing: Lumen.Spacing.sm) {
                    SoundPicker(label: "重拍音色", selection: $vm.accentSound)
                    SoundPicker(label: "弱拍音色", selection: $vm.normalSound)
                }

                Spacer()

                // 播放/停止大按钮（全宽 72dp，播放中 tune/off 红）
                Button {
                    vm.togglePlay()
                } label: {
                    Text(vm.playing ? "停止" : "播放")
                        .font(.title3.bold())
                        .foregroundStyle(.white)
                        .frame(maxWidth: .infinity)
                        .frame(height: 72)
                        .background(
                            vm.playing ? palette.tuneOff : palette.accent,
                            in: RoundedRectangle(cornerRadius: 28)
                        )
                }
            }
            .padding(Lumen.Spacing.page)
        }
    }
}

/// 拍点（重拍实心/普通半透/静音空心）。
struct BeatDot: View {
    @Environment(\.lumen) private var palette
    var accent: TickAccent
    var current: Bool
    var onClick: () -> Void

    var body: some View {
        Button(action: onClick) {
            Group {
                switch accent {
                case .accent:
                    Circle().fill(current ? palette.accent : palette.inkPrimary)
                case .normal:
                    Circle().fill((current ? palette.accent : palette.inkPrimary).opacity(0.4))
                case .muted:
                    Circle().stroke(current ? palette.accent : palette.inkSecondary, lineWidth: 1.5)
                }
            }
            .frame(width: current ? 18 : 12, height: current ? 18 : 12)
        }
    }
}

/// 步进按钮。
struct StepButton: View {
    @Environment(\.lumen) private var palette
    var text: String
    var action: () -> Void

    init(_ text: String, action: @escaping () -> Void) {
        self.text = text
        self.action = action
    }

    var body: some View {
        Button(action: action) {
            Text(text)
                .font(Lumen.label)
                .foregroundStyle(palette.accent)
                .padding(.horizontal, 12)
                .padding(.vertical, 8)
                .background(palette.bgSurface, in: Capsule())
                .overlay(Capsule().stroke(palette.lineSubtle, lineWidth: 1))
        }
    }
}

/// 音色选择器。
struct SoundPicker: View {
    @Environment(\.lumen) private var palette
    var label: String
    @Binding var selection: TickSoundKind

    var body: some View {
        Menu {
            ForEach(TickSoundKind.allCases, id: \.self) { kind in
                Button(kind.label) { selection = kind }
            }
        } label: {
            VStack(alignment: .leading, spacing: 4) {
                Text(label)
                    .font(Lumen.caption)
                    .foregroundStyle(palette.inkSecondary)
                HStack {
                    Text(selection.label)
                        .font(Lumen.label)
                        .foregroundStyle(palette.inkPrimary)
                    Image(systemName: "chevron.down")
                        .font(.caption2)
                        .foregroundStyle(palette.inkSecondary)
                }
            }
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(.horizontal, 14)
            .padding(.vertical, 10)
            .background(palette.bgSurface, in: RoundedRectangle(cornerRadius: 12))
        }
    }
}

File: android/app/src/main/kotlin/com/liuchong/tunar/ui/metronome/MetronomeScreen.kt
package com.liuchong.tunar.ui.metronome

import android.Manifest
import android.content.pm.PackageManager
import android.os.Build
import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.gestures.detectVerticalDragGestures
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.border
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ExposedDropdownMenuBox
import androidx.compose.material3.ExposedDropdownMenuDefaults
import androidx.compose.material3.FilterChip
import androidx.compose.material3.LinearProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.MenuAnchorType
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Slider
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableDoubleStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.core.content.ContextCompat
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import androidx.lifecycle.createSavedStateHandle
import androidx.lifecycle.viewmodel.compose.viewModel
import com.liuchong.tunar.audio.AudioTrackMetronomePlayer
import com.liuchong.tunar.audio.MetronomeService
import com.liuchong.tunar.ui.common.AuroraBackground
import com.liuchong.tunar.corebinding.TickSoundKind
import com.liuchong.tunar.corebinding.TickSounds
import com.liuchong.tunar.corebinding.uniffiMetronomeFactory
import com.liuchong.tunar.ui.theme.LocalLumenColors
import com.liuchong.tunar.ui.theme.TunarTypography
import uniffi.tunar_core.TickAccent

/** 节拍器面板（spec-ui §3）。 */
@Composable
fun MetronomeScreen(
    viewModel: MetronomeViewModel = viewModel(initializer = {
        MetronomeViewModel(
            engineFactory = uniffiMetronomeFactory,
            player = AudioTrackMetronomePlayer(),
            sounds = TickSounds.buildAll(),
            savedState = createSavedStateHandle(),
        )
    }),
) {
    val state by viewModel.uiState.collectAsStateWithLifecycle()
    val context = LocalContext.current

    // 通知栏停止按钮 → 停止播放
    DisposableEffect(Unit) {
        MetronomeService.onStopRequested = { viewModel.pause() }
        onDispose { MetronomeService.onStopRequested = null }
    }

    // Android 13+ 通知权限（拒绝时播放不受影响，仅通知栏不显示）
    val notifLauncher = rememberLauncherForActivityResult(
        ActivityResultContracts.RequestPermission(),
    ) { }

    fun onPlayToggle() {
        val starting = viewModel.togglePlay()
        if (starting) {
            if (Build.VERSION.SDK_INT >= 33 &&
                ContextCompat.checkSelfPermission(context, Manifest.permission.POST_NOTIFICATIONS) !=
                PackageManager.PERMISSION_GRANTED
            ) {
                notifLauncher.launch(Manifest.permission.POST_NOTIFICATIONS)
            }
            MetronomeService.start(context, viewModel.uiState.value.bpm)
        } else {
            MetronomeService.stop(context)
        }
    }

    AuroraBackground(tuneCents = null) {
    Column(
        modifier = Modifier.fillMaxSize().padding(16.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        // BPM 环 + 垂直拨轮（design-system §6.5）
        var dragAccum by remember { mutableDoubleStateOf(0.0) }
        Box(
            modifier = Modifier.pointerInput(Unit) {
                detectVerticalDragGestures { change, dragAmount ->
                    change.consume()
                    dragAccum += -dragAmount
                    while (dragAccum >= 16.0) {
                        viewModel.adjustBpm(1)
                        dragAccum -= 16.0
                    }
                    while (dragAccum <= -16.0) {
                        viewModel.adjustBpm(-1)
                        dragAccum += 16.0
                    }
                }
            },
        ) {
            BpmRing(bpm = state.bpm) {
                Text(
                    text = "BPM（上下拖动调节）",
                    style = MaterialTheme.typography.labelMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }
        }

        // 摆锤（播放时摆动，停止归中）
        Pendulum(
            playing = state.playing,
            bpm = state.bpm,
            beatUnit = state.beatUnit,
        )

        // 滑杆 + 步进 + tap
        Row(verticalAlignment = Alignment.CenterVertically, modifier = Modifier.fillMaxWidth()) {
            OutlinedButton(onClick = { viewModel.adjustBpm(-5) }) { Text("-5") }
            OutlinedButton(onClick = { viewModel.adjustBpm(-1) }) { Text("-1") }
            Slider(
                value = state.bpm.toFloat(),
                onValueChange = { viewModel.setBpm(it.toDouble()) },
                valueRange = 30f..250f,
                modifier = Modifier.weight(1f).padding(horizontal = 8.dp),
            )
            OutlinedButton(onClick = { viewModel.adjustBpm(1) }) { Text("+1") }
            OutlinedButton(onClick = { viewModel.adjustBpm(5) }) { Text("+5") }
        }
        OutlinedButton(onClick = { viewModel.tap() }) { Text("TAP 测速") }

        Spacer(modifier = Modifier.height(8.dp))

        // 拍号选择
        Row(
            modifier = Modifier.fillMaxWidth().horizontalScroll(rememberScrollState()),
            horizontalArrangement = Arrangement.spacedBy(6.dp),
        ) {
            COMMON_TIME_SIGNATURES.forEach { (beats, unit) ->
                FilterChip(
                    selected = state.beatsPerBar == beats && state.beatUnit == unit,
                    onClick = { viewModel.setTimeSignature(beats, unit) },
                    label = { Text("$beats/$unit") },
                )
            }
        }

        Spacer(modifier = Modifier.height(8.dp))

        // 重音圆点行（点击循环 重拍/普通/静音；当前拍放大高亮）
        Row(
            horizontalArrangement = Arrangement.spacedBy(12.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            state.accents.forEachIndexed { i, accent ->
                val isCurrent = state.playing && state.currentBeat == i
                val colors = LocalLumenColors.current
                // 重拍实心 / 普通半透 / 静音空心（形状+填充双重编码）
                val dotModifier = Modifier
                    .size(if (isCurrent) 18.dp else 12.dp)
                    .clip(CircleShape)
                    .clickable { viewModel.cycleAccent(i) }
                when (accent) {
                    TickAccent.ACCENT -> Box(
                        dotModifier.background(
                            if (isCurrent) colors.accent else colors.inkPrimary,
                        ),
                    )
                    TickAccent.NORMAL -> Box(
                        dotModifier.background(
                            (if (isCurrent) colors.accent else colors.inkPrimary)
                                .copy(alpha = 0.4f),
                        ),
                    )
                    TickAccent.MUTED -> Box(
                        dotModifier.border(
                            1.5.dp,
                            if (isCurrent) colors.accent else colors.inkSecondary,
                            CircleShape,
                        ),
                    )
                }
            }
        }
        // 小节进度
        LinearProgressIndicator(
            progress = {
                if (state.playing && state.currentBeat >= 0) {
                    (state.currentBeat + 1).toFloat() / state.beatsPerBar
                } else {
                    0f
                }
            },
            modifier = Modifier.fillMaxWidth().padding(vertical = 8.dp),
        )

        // 音色选择（重拍/弱拍）
        Row(modifier = Modifier.fillMaxWidth()) {
            SoundDropdown(
                label = "重拍音色",
                value = state.accentSound,
                onSelect = viewModel::setAccentSound,
                modifier = Modifier.weight(1f),
            )
            Spacer(modifier = Modifier.width(8.dp))
            SoundDropdown(
                label = "弱拍音色",
                value = state.normalSound,
                onSelect = viewModel::setNormalSound,
                modifier = Modifier.weight(1f),
            )
        }

        Spacer(modifier = Modifier.weight(1f))

        // 播放/停止大按钮（design-system §6.4：全宽 72dp、圆角 28dp，播放中 tune/off 红）
        Button(
            onClick = { onPlayToggle() },
            modifier = Modifier.fillMaxWidth().height(72.dp),
            shape = RoundedCornerShape(28.dp),
            colors = if (state.playing) {
                ButtonDefaults.buttonColors(containerColor = LocalLumenColors.current.tuneOff)
            } else {
                ButtonDefaults.buttonColors()
            },
        ) {
            Text(
                text = if (state.playing) "停止" else "播放",
                style = MaterialTheme.typography.titleLarge,
            )
        }
    }
    }
}

/** 音色下拉。 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
private fun SoundDropdown(
    label: String,
    value: TickSoundKind,
    onSelect: (TickSoundKind) -> Unit,
    modifier: Modifier = Modifier,
) {
    var expanded by remember { mutableStateOf(false) }
    ExposedDropdownMenuBox(
        expanded = expanded,
        onExpandedChange = { expanded = it },
        modifier = modifier,
    ) {
        OutlinedTextField(
            value = value.label,
            onValueChange = {},
            readOnly = true,
            label = { Text(label) },
            trailingIcon = { ExposedDropdownMenuDefaults.TrailingIcon(expanded = expanded) },
            modifier = Modifier
                .menuAnchor(MenuAnchorType.PrimaryNotEditable)
                .fillMaxWidth(),
        )
        ExposedDropdownMenu(expanded = expanded, onDismissRequest = { expanded = false }) {
            TickSoundKind.entries.forEach { kind ->
                DropdownMenuItem(
                    text = { Text(kind.label) },
                    onClick = {
                        onSelect(kind)
                        expanded = false
                    },
                )
            }
        }
    }
}
